pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+
+	"github.com/spiffe/spire/pkg/agent/api/rpccontext"
+	workload_attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/endpoints"
+	"github.com/spiffe/spire/pkg/agent/manager"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/proto/spire/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterService registers privileged service on provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type Attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log             logrus.FieldLogger
+	Manager         manager.Manager
+	Attestor        workload_attestor.Attestor
+	AuthorizedUsers []string
+}
+
+func New(config Config) *Service {
+	authorizedUsers := map[string]bool{}
+
+	for _, user := range config.AuthorizedUsers {
+		authorizedUsers[user] = true
+	}
+
+	return &Service{
+		manager:         config.Manager,
+		attestor:        endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		authorizedUsers: authorizedUsers,
+	}
+}
+
+// Service implements the privileged server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor Attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	selectors, err := api.SelectorsFromProto(req.Selectors)
+	if err != nil {
+		return err
```

Comment:
We need a grpc code here, and we should also log the error, something like the following

```suggestion
		log.WithError(err).Error("Invalid argument; could not parse provided selectors")
		return status.Error(codes.InvalidArgument, "could not parse provided selectors")
```